README.md
# Awesome Bottom Navigation

Custom bottom navigation bar using flutter. UI inspiration from [Dribble](https://dribbble.com/shots/6005981-Tab-Bar-Animation-nr-2).

![Preview](https://media.giphy.com/media/bDaJZ80hHPRyJsBGfB/giphy.gif)

## Basic Usage

```dart
bottomNavigationBar: AwesomeBottomNav(
    icons: [
        Icons.home_outlined,
        Icons.shopping_cart_outlined,
        Icons.category_outlined,
        Icons.account_circle_outlined,
    ],
    highlightedIcons: [
        Icons.home,
        Icons.shopping_cart,
        Icons.category,
        Icons.account_circle,
    ],
    onTapped: (int value) {
        setState(() {
            selectedIndex = value;
        });
    },
    bodyBgColor: _bgColor,
    highlightColor: Color(0xFFFF9944),
    navFgColor: Colors.grey.withOpacity(0.5),
    navBgColor: Colors.white,
),
```


CHANGELOG.md
## [0.0.1] - 4th March 2021 - Initial beta

* TODO: Polishing


## [0.0.2] - 5th March 2021 - Stable

* Added better description.
* Refactored to hide inner classes
* TODO: Add dart documentation on API

## [0.0.3] - 5th March 2021 - Stable

* Added dart documentation on API


test/awesome_bottom_navigation_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:awesome_bottom_navigation/awesome_bottom_navigation.dart';

void main() {

}


lib/src/clipper.dart
import 'consts.dart';
import 'package:flutter/material.dart';
import 'package:vector_math/vector_math_64.dart';

class BottomNavClipper extends CustomClipper<Path> {
  final int numberOfTabs;
  final double notchHeight = kNavSize - 8;
  final topPaddingFactor = 0.0;
  final BuildContext context;

  final double animatedIndex;
  double paddingW;

  BottomNavClipper({
    @required this.numberOfTabs,
    @required this.animatedIndex,
    @required this.context,
  }) {
    final size = MediaQuery.of(context).size;
    final totalPadding = size.width - (kNavSize * numberOfTabs);
    paddingW = totalPadding / (numberOfTabs + 1);
  }

  @override
  Path getClip(Size size) {
    var path = Path();
    path.moveTo(0, 0);
    final sectionWidth = (size.width - paddingW) / numberOfTabs;
    final curveControlOffset = sectionWidth * 0.5;

    final topPadding = topPaddingFactor * size.height;
    path.lineTo(
        (animatedIndex * sectionWidth + paddingW / 2) - curveControlOffset, 0);

    final firstControlPoint =
        Offset((animatedIndex * sectionWidth + paddingW / 2), 0);

    final secondControlPoint =
        Offset((animatedIndex * sectionWidth + paddingW / 2), notchHeight);
    final secondEndPoint = Offset(
        (animatedIndex * sectionWidth + paddingW / 2) + curveControlOffset,
        notchHeight);

    path.cubicTo(
        firstControlPoint.dx,
        firstControlPoint.dy,
        secondControlPoint.dx,
        secondControlPoint.dy,
        secondEndPoint.dx,
        secondEndPoint.dy);

    path.lineTo(
        ((animatedIndex + 1) * sectionWidth + paddingW / 2) -
            curveControlOffset,
        notchHeight);
    final thirdControlPoint = Offset(
        ((animatedIndex + 1) * sectionWidth + paddingW / 2), notchHeight);

    final fourthControlPoint =
        Offset(((animatedIndex + 1) * sectionWidth + paddingW / 2), 0);
    final fourthEndPoint = Offset(
        ((animatedIndex + 1) * sectionWidth + paddingW / 2) +
            curveControlOffset,
        0);

    path.cubicTo(
        thirdControlPoint.dx,
        thirdControlPoint.dy,
        fourthControlPoint.dx,
        fourthControlPoint.dy,
        fourthEndPoint.dx,
        fourthEndPoint.dy);
    path.lineTo(size.width, 0);

    path =
        path.transform(Matrix4.translation(Vector3(0, topPadding, 0)).storage);

    path.lineTo(size.width, size.height);
    path.lineTo(0, size.height);
    path.close();
    return path;
  }

  @override
  bool shouldReclip(covariant BottomNavClipper oldClipper) {
    return oldClipper.animatedIndex != animatedIndex;
  }
}


example/lib/main.dart
import 'package:awesome_bottom_navigation/awesome_bottom_navigation.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Bottom Nav Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: ExampleHomePage(),
    );
  }
}

class ExampleHomePage extends StatefulWidget {
  @override
  _ExampleHomePageState createState() => _ExampleHomePageState();
}

class _ExampleHomePageState extends State<ExampleHomePage> {
  int selectedIndex = 0;
  final _bgColor = Color(0xFFF6F6F6);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        color: _bgColor,
        child: Center(
          child: Text("Selected Page: $selectedIndex"),
        ),
      ),
      bottomNavigationBar: AwesomeBottomNav(
        icons: [
          Icons.home_outlined,
          Icons.shopping_cart_outlined,
          Icons.category_outlined,
          Icons.account_circle_outlined,
          // Icons.settings_outlined,
        ],
        highlightedIcons: [
          Icons.home,
          Icons.shopping_cart,
          Icons.category,
          Icons.account_circle,
          // Icons.settings,
        ],
        onTapped: (int value) {
          setState(() {
            selectedIndex = value;
          });
        },
        bodyBgColor: _bgColor,
        highlightColor: Color(0xFFFF9944),
        navFgColor: Colors.grey.withOpacity(0.5),
        navBgColor: Colors.white,
      ),
    );
  }
}


lib/src/shadow_painter.dart
import 'consts.dart';
import 'package:flutter/material.dart';
import 'package:vector_math/vector_math_64.dart' as vector;

class BottomNavShadowPainter extends CustomPainter {
  final numberOfTabs;
  final double notchHeight = kNavSize - 8;
  final topPaddingFactor = 0.0;
  final BuildContext context;

  final double animatedIndex;
  double paddingW;

  BottomNavShadowPainter({
    @required this.numberOfTabs,
    @required this.animatedIndex,
    @required this.context,
  }) {
    final size = MediaQuery.of(context).size;
    final totalPadding = size.width - (kNavSize * numberOfTabs);
    paddingW = totalPadding / (numberOfTabs + 1);
  }

  @override
  void paint(Canvas canvas, Size size) {
    var path = Path();
    path.moveTo(0, 0);
    final sectionWidth = (size.width - paddingW) / numberOfTabs;
    final curveControlOffset = sectionWidth * 0.5;

    final topPadding = topPaddingFactor * size.height;
    path.lineTo(
        (animatedIndex * sectionWidth + paddingW / 2) - curveControlOffset, 0);

    final firstControlPoint =
        Offset((animatedIndex * sectionWidth + paddingW / 2), 0);

    final secondControlPoint =
        Offset((animatedIndex * sectionWidth + paddingW / 2), notchHeight);
    final secondEndPoint = Offset(
        (animatedIndex * sectionWidth + paddingW / 2) + curveControlOffset,
        notchHeight);

    path.cubicTo(
        firstControlPoint.dx,
        firstControlPoint.dy,
        secondControlPoint.dx,
        secondControlPoint.dy,
        secondEndPoint.dx,
        secondEndPoint.dy);

    path.lineTo(
        ((animatedIndex + 1) * sectionWidth + paddingW / 2) -
            curveControlOffset,
        notchHeight);
    final thirdControlPoint = Offset(
        ((animatedIndex + 1) * sectionWidth + paddingW / 2), notchHeight);

    final fourthControlPoint =
        Offset(((animatedIndex + 1) * sectionWidth + paddingW / 2), 0);
    final fourthEndPoint = Offset(
        ((animatedIndex + 1) * sectionWidth + paddingW / 2) +
            curveControlOffset,
        0);

    path.cubicTo(
        thirdControlPoint.dx,
        thirdControlPoint.dy,
        fourthControlPoint.dx,
        fourthControlPoint.dy,
        fourthEndPoint.dx,
        fourthEndPoint.dy);
    path.lineTo(size.width, 0);

    path = path.transform(
        Matrix4.translation(vector.Vector3(0, topPadding, 0)).storage);

    path.lineTo(size.width, size.height);
    path.lineTo(0, size.height);
    path.close();
    canvas.drawShadow(path, Colors.black, 14, false);
  }

  @override
  bool shouldRepaint(covariant BottomNavShadowPainter oldDelegate) {
    return oldDelegate.animatedIndex != animatedIndex;
  }
}


lib/src/bottom_nav.dart
import 'package:flutter/material.dart';
import 'clipper.dart';
import 'consts.dart';
import 'shadow_painter.dart';

/// A Bottom Navigation Bar that comes with beautiful animation stylish curves.
///
/// The design is inspired from a [dribble post](https://dribbble.com/shots/6005981-Tab-Bar-Animation-nr-2)
/// {@tool snippet}
///
/// This is a basic usage in a Scaffold
/// ```dart
/// Scaffold(
///   body: Container(
///     color: _bgColor,
///     child: Center(
///       child: Text("Selected Page: $selectedIndex"),
///     ),
///   ),
///   bottomNavigationBar: AwesomeBottomNav(
///     icons: [
///       Icons.home_outlined,
///       Icons.shopping_cart_outlined,
///       Icons.category_outlined,
///       Icons.account_circle_outlined,
///       // Icons.settings_outlined,
///     ],
///     highlightedIcons: [
///       Icons.home,
///       Icons.shopping_cart,
///       Icons.category,
///       Icons.account_circle,
///       // Icons.settings,
///     ],
///     onTapped: (int value) {
///       setState(() {
///         selectedIndex = value;
///       });
///     },
///     bodyBgColor: _bgColor,
///     highlightColor: Color(0xFFFF9944),
///     navFgColor: Colors.grey.withOpacity(0.5),
///     navBgColor: Colors.white,
///   ),
/// );
/// ```
class AwesomeBottomNav extends StatefulWidget {
  /// List of icons of type [IconData] to be shown in the bottom navigation in unselected state.
  final List<IconData> icons;

  /// List of icons of type [IconData] to be shown in the bottom navigation in selected state.
  final List<IconData> highlightedIcons;

  /// Color of highlight
  final Color highlightColor;

  final Color bodyBgColor;
  final Color navBgColor;
  final Color navFgColor;
  final ValueSetter<int> onTapped;

  const AwesomeBottomNav({
    Key key,
    @required this.icons,
    @required this.highlightedIcons,
    @required this.highlightColor,
    @required this.onTapped,
    @required this.bodyBgColor,
    @required this.navBgColor,
    @required this.navFgColor,
  }) : super(key: key);

  @override
  _AwesomeBottomNavState createState() => _AwesomeBottomNavState();
}

class _AwesomeBottomNavState extends State<AwesomeBottomNav>
    with SingleTickerProviderStateMixin {
  int _selectedIndex = 0;
  AnimationController _animationController;
  Animation<double> _posXAnimation;
  Animation<double> _sinkAnimation;
  Animation<double> _riseAnimation;

  @override
  void initState() {
    super.initState();
    _animationController = AnimationController(
      duration: Duration(milliseconds: 300),
      vsync: this,
    );
    _posXAnimation =
        Tween<double>(begin: 0.0, end: (_selectedIndex + 1) * 1.0).animate(
      CurvedAnimation(
        parent: _animationController,
        curve: Curves.easeInOutSine,
      ),
    );
    _sinkAnimation =
        Tween(begin: -20.0, end: kNavSize.toDouble()).animate(CurvedAnimation(
      parent: _animationController,
      curve: Interval(0.0, 0.25, curve: Curves.easeInOutSine),
    ));
    _riseAnimation =
        Tween(end: -20.0, begin: kNavSize.toDouble()).animate(CurvedAnimation(
      parent: _animationController,
      curve: Interval(0.75, 1.0, curve: Curves.easeInOutSine),
    ));
  }

  void _tapped(int index) {
    if (_selectedIndex == index) return;
    _animationController.reset();
    _posXAnimation =
        Tween<double>(begin: _selectedIndex * 1.0, end: index * 1.0)
            .animate(CurvedAnimation(
      parent: _animationController,
      curve: Curves.easeInOutSine,
    ));
    _animationController.forward();
    setState(() {
      _selectedIndex = index;
    });
    widget.onTapped(index);
  }

  double _getMainCircleTop() {
    if (!_animationController.isAnimating) {
      return -20;
    }

    if (_animationController.value < 0.5) {
      return _sinkAnimation.value;
    } else {
      return _riseAnimation.value;
    }
  }

  double _getMainCircleLeft(Size size) {
    final totalPadding = size.width - (kNavSize * widget.icons.length);
    final singlePadding = totalPadding / (widget.icons.length + 1);
    return ((_animationController.isAnimating
                ? _posXAnimation.value
                : _selectedIndex) *
            kNavSize) +
        (singlePadding *
            ((_animationController.isAnimating
                    ? _posXAnimation.value
                    : _selectedIndex) +
                1));
  }

  Widget _buildCircle(Size _size) {
    return AnimatedBuilder(
      animation: _animationController,
      builder: (context, child) => Positioned(
        top: _getMainCircleTop(),
        left: _getMainCircleLeft(_size),
        child: SizedBox(
          height: kNavSize,
          width: kNavSize,
          child: Material(
            color: widget.highlightColor,
            clipBehavior: Clip.antiAlias,
            type: MaterialType.circle,
            elevation: 2,
            child: Icon(
              widget.highlightedIcons[_selectedIndex],
              color: Colors.white,
              size: 30,
            ),
          ),
        ),
      ),
    );
  }

  Widget _buildBar(BuildContext context) {
    return AnimatedBuilder(
      animation: _animationController,
      builder: (context, child) => CustomPaint(
        painter: BottomNavShadowPainter(
          context: context,
          animatedIndex: _animationController.isAnimating
              ? _posXAnimation.value
              : _selectedIndex * 1.0,
          numberOfTabs: widget.icons.length,
        ),
        child: ClipPath(
          clipBehavior: Clip.antiAlias,
          clipper: BottomNavClipper(
            context: context,
            animatedIndex: _animationController.isAnimating
                ? _posXAnimation.value
                : _selectedIndex * 1.0,
            numberOfTabs: widget.icons.length,
          ),
          child: Container(
            height: kNavSize + MediaQuery.of(context).padding.bottom,
            child: Material(
              color: widget.navBgColor,
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                crossAxisAlignment: CrossAxisAlignment.start,
                children: widget.icons
                    .map<Widget>((icon) => InkWell(
                          borderRadius: BorderRadius.circular(kNavSize),
                          onTap: () {
                            _tapped(widget.icons.indexOf(icon));
                          },
                          child: Container(
                            height: 64,
                            width: 56,
                            child: Center(
                              child: Icon(
                                icon,
                                size: 30,
                                color: widget.navFgColor,
                              ),
                            ),
                          ),
                        ))
                    .toList(),
              ),
            ),
          ),
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    final _size = MediaQuery.of(context).size;
    return Container(
      color: widget.bodyBgColor,
      child: Stack(
        clipBehavior: Clip.none,
        children: [
          _buildCircle(_size),
          _buildBar(context),
        ],
      ),
    );
  }
}
